File: internal/tmpl/tmpl_test.go
package tmpl

import (
	"os"
	"path/filepath"
	"testing"
)

func contains(list []string, want string) bool {
	for _, s := range list {
		if s == want {
			return true
		}
	}
	return false
}

func TestExistsBuiltin(t *testing.T) {
	if !Exists("", "headers", "class", "source.h.tmpl") {
		t.Error("Exists reported built-in headers/class/source.h.tmpl missing")
	}
	if Exists("", "headers", "class", "nope.tmpl") {
		t.Error("Exists reported a nonexistent template present")
	}
}

func TestReadBuiltin(t *testing.T) {
	b, err := Read("", "headers", "class", "source.h.tmpl")
	if err != nil {
		t.Fatalf("Read built-in returned error: %v", err)
	}
	if len(b) == 0 {
		t.Error("Read returned empty content for a built-in template")
	}
}

func TestIsCompiled(t *testing.T) {
	// class ships a .h/.cpp pair -> compiled.
	if !IsCompiled("", "headers", "class") {
		t.Error("headers/class should be compiled (has source.h + source.cpp)")
	}
	// templated-class is header-only (.hpp) -> not compiled.
	if IsCompiled("", "headers", "templated-class") {
		t.Error("headers/templated-class should not be compiled (header-only)")
	}
	// modules kinds are never compiled in the headers sense.
	if IsCompiled("", "modules", "class") {
		t.Error("modules/class should not report as compiled")
	}
}

func TestKindsExcludesSpecialDirs(t *testing.T) {
	kinds := Kinds("", "headers")
	for _, want := range []string{"class", "interface", "templated-class"} {
		if !contains(kinds, want) {
			t.Errorf("Kinds(headers) missing component kind %q; got %v", want, kinds)
		}
	}
	for _, excluded := range []string{"app", "test", "project"} {
		if contains(kinds, excluded) {
			t.Errorf("Kinds(headers) should exclude special dir %q; got %v", excluded, kinds)
		}
	}
}

func TestKindsModules(t *testing.T) {
	if kinds := Kinds("", "modules"); !contains(kinds, "class") {
		t.Errorf("Kinds(modules) missing class (has source.cppm.tmpl); got %v", kinds)
	}
}

func TestBuiltinKindsIncludesSpecialDirs(t *testing.T) {
	kinds := BuiltinKinds("headers")
	for _, want := range []string{"app", "test", "project", "class"} {
		if !contains(kinds, want) {
			t.Errorf("BuiltinKinds(headers) missing %q; got %v", want, kinds)
		}
	}
}

func TestProjectOverrideWins(t *testing.T) {
	root := t.TempDir()
	local := filepath.Join(root, ProjectTemplateDir, "class", "source.h.tmpl")
	if err := os.MkdirAll(filepath.Dir(local), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(local, []byte("OVERRIDE"), 0o644); err != nil {
		t.Fatal(err)
	}

	b, err := Read(root, "headers", "class", "source.h.tmpl")
	if err != nil {
		t.Fatalf("Read with override returned error: %v", err)
	}
	if string(b) != "OVERRIDE" {
		t.Errorf("Read = %q, want project override %q", b, "OVERRIDE")
	}
	if !Exists(root, "headers", "class", "source.h.tmpl") {
		t.Error("Exists should see the project override")
	}
}

func TestProjectOverrideAddsNewKind(t *testing.T) {
	root := t.TempDir()
	// A brand-new headers component kind, defined only in the project.
	local := filepath.Join(root, ProjectTemplateDir, "widget", "source.hpp.tmpl")
	if err := os.MkdirAll(filepath.Dir(local), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(local, []byte("// widget"), 0o644); err != nil {
		t.Fatal(err)
	}

	if kinds := Kinds(root, "headers"); !contains(kinds, "widget") {
		t.Errorf("Kinds should include project-local kind 'widget'; got %v", kinds)
	}
}

func TestCopyBuiltin(t *testing.T) {
	dst := filepath.Join(t.TempDir(), "copied")
	if err := CopyBuiltin("headers", "class", dst); err != nil {
		t.Fatalf("CopyBuiltin returned error: %v", err)
	}
	for _, name := range []string{"source.h.tmpl", "source.cpp.tmpl", "CMakeLists.txt.tmpl"} {
		if _, err := os.Stat(filepath.Join(dst, name)); err != nil {
			t.Errorf("CopyBuiltin did not write %s: %v", name, err)
		}
	}
}

File: src/tests/tmpl_test.cpp

#include <catch2/catch_test_macros.hpp>

#include <unistd.h>

#include <algorithm>
#include <expected>
#include <filesystem>
#include <format>
#include <string>
#include <vector>

#include "TempDir.hpp"

import cup.tmpl;

namespace {

using cup::test::TempDir;

const std::filesystem::path kBuiltinsOnly;

[[nodiscard]] bool contains(const std::vector<std::string>& list, std::string_view want) {
    return std::ranges::find(list, want) != list.end();
}

}

TEST_CASE("exists finds built-in templates and rejects absent ones", "[tmpl][exists]") {
    REQUIRE(cup::tmpl::exists(kBuiltinsOnly, "headers", "class", "source.h.tmpl"));
    REQUIRE_FALSE(cup::tmpl::exists(kBuiltinsOnly, "headers", "class", "nope.tmpl"));
}

TEST_CASE("read returns built-in template content", "[tmpl][read]") {
    const auto got = cup::tmpl::read(kBuiltinsOnly, "headers", "class", "source.h.tmpl");
    REQUIRE(got.has_value());
    REQUIRE_FALSE(got->empty());
}

TEST_CASE("read reports a missing template", "[tmpl][read]") {
    const auto got = cup::tmpl::read(kBuiltinsOnly, "headers", "class", "nope.tmpl");
    REQUIRE_FALSE(got.has_value());
}

TEST_CASE("is_compiled distinguishes .h/.cpp pairs from header-only kinds", "[tmpl][compiled]") {
    REQUIRE(cup::tmpl::is_compiled(kBuiltinsOnly, "headers", "class"));
    REQUIRE_FALSE(cup::tmpl::is_compiled(kBuiltinsOnly, "headers", "templated-class"));
    REQUIRE_FALSE(cup::tmpl::is_compiled(kBuiltinsOnly, "modules", "class"));
}

TEST_CASE("kinds lists component kinds and excludes the special dirs", "[tmpl][kinds]") {
    const auto kinds = cup::tmpl::kinds(kBuiltinsOnly, "headers");
    for (const auto* want : {"class", "interface", "templated-class"}) {
        INFO("expected component kind: " << want);
        REQUIRE(contains(kinds, want));
    }
    for (const auto* excluded : {"app", "test", "project"}) {
        INFO("special dir that must not be offered: " << excluded);
        REQUIRE_FALSE(contains(kinds, excluded));
    }
}

TEST_CASE("kinds finds modules-family kinds by their interface unit", "[tmpl][kinds]") {
    REQUIRE(contains(cup::tmpl::kinds(kBuiltinsOnly, "modules"), "class"));
}

TEST_CASE("builtin_kinds includes the special dirs", "[tmpl][kinds]") {
    const auto kinds = cup::tmpl::builtin_kinds("headers");
    for (const auto* want : {"app", "test", "project", "class"}) {
        INFO("expected built-in dir: " << want);
        REQUIRE(contains(kinds, want));
    }
}

TEST_CASE("a project-local template overrides the built-in", "[tmpl][override]") {
    const TempDir root;
    root.write(std::filesystem::path(cup::tmpl::kProjectTemplateDir) / "class" / "source.h.tmpl",
               "OVERRIDE");

    const auto got = cup::tmpl::read(root, "headers", "class", "source.h.tmpl");
    REQUIRE(got.has_value());
    REQUIRE(*got == "OVERRIDE");
    REQUIRE(cup::tmpl::exists(root, "headers", "class", "source.h.tmpl"));
}

TEST_CASE("a project-local directory adds a new kind", "[tmpl][override][kinds]") {
    const TempDir root;
    root.write(std::filesystem::path(cup::tmpl::kProjectTemplateDir) / "widget" / "source.hpp.tmpl",
               "// widget");

    REQUIRE(contains(cup::tmpl::kinds(root, "headers"), "widget"));
}

TEST_CASE("kinds ignores project template entries that are not usable kinds",
          "[tmpl][kinds][override]") {
    const TempDir root;
    const std::filesystem::path templates(cup::tmpl::kProjectTemplateDir);
    root.write(templates / "notes.md", "not a kind");
    root.write(templates / "empty-kind" / "README.md", "no source here");

    const auto kinds = cup::tmpl::kinds(root, "headers");
    REQUIRE_FALSE(contains(kinds, "notes.md"));
    REQUIRE_FALSE(contains(kinds, "empty-kind"));
    REQUIRE(contains(kinds, "class"));
}

TEST_CASE("is_compiled needs both halves of the declaration/definition pair",
          "[tmpl][compiled][override]") {
    const TempDir root;
    root.write(std::filesystem::path(cup::tmpl::kProjectTemplateDir) / "halfclass" /
                   "source.h.tmpl",
               "// declaration only");

    REQUIRE(cup::tmpl::exists(root, "headers", "halfclass", "source.h.tmpl"));
    REQUIRE_FALSE(cup::tmpl::exists(root, "headers", "halfclass", "source.cpp.tmpl"));
    REQUIRE_FALSE(cup::tmpl::is_compiled(root, "headers", "halfclass"));
    REQUIRE_FALSE(contains(cup::tmpl::kinds(root, "headers"), "halfclass"));
}

TEST_CASE("copy_builtin writes every file of a built-in kind", "[tmpl][copy]") {
    const TempDir tmp;
    const std::filesystem::path dst = tmp.path() / "copied";

    REQUIRE(cup::tmpl::copy_builtin("headers", "class", dst).has_value());
    for (const auto* name : {"source.h.tmpl", "source.cpp.tmpl", "CMakeLists.txt.tmpl"}) {
        INFO("expected copied file: " << name);
        REQUIRE(std::filesystem::exists(dst / name));
    }
}

TEST_CASE("copy_builtin reports an unknown kind", "[tmpl][copy]") {
    const TempDir tmp;
    REQUIRE_FALSE(cup::tmpl::copy_builtin("headers", "ghost", tmp.path() / "copied").has_value());
}

TEST_CASE("an unreadable override falls through to the built-in", "[tmpl][override]") {
    if (::geteuid() == 0) {
        SKIP("root ignores the permission bits, so the override stays readable");
    }
    const TempDir root;
    const std::filesystem::path override_file =
        std::filesystem::path(cup::tmpl::kProjectTemplateDir) / "class" / "source.h.tmpl";
    root.write(override_file, "OVERRIDE");
    std::filesystem::permissions(root.path() / override_file, std::filesystem::perms::none);

    REQUIRE(cup::tmpl::exists(root, "headers", "class", "source.h.tmpl"));

    const auto got = cup::tmpl::read(root, "headers", "class", "source.h.tmpl");
    REQUIRE(got.has_value());
    REQUIRE(*got != "OVERRIDE");
    REQUIRE(*got == cup::tmpl::read(kBuiltinsOnly, "headers", "class", "source.h.tmpl").value());
}

TEST_CASE("copy_builtin reports a destination it cannot create", "[tmpl][copy]") {
    const TempDir tmp;
    tmp.write("blocker", "not a directory");

    const auto copied = cup::tmpl::copy_builtin("headers", "class", tmp.path() / "blocker" / "sub");
    REQUIRE_FALSE(copied.has_value());
    REQUIRE(copied.error().message().starts_with("creating "));
}

TEST_CASE("copy_builtin reports a file it cannot write", "[tmpl][copy]") {
    const TempDir tmp;
    const std::filesystem::path dst = tmp.path() / "copied";
    std::filesystem::create_directories(dst / "source.h.tmpl");

    const auto copied = cup::tmpl::copy_builtin("headers", "class", dst);
    REQUIRE_FALSE(copied.has_value());
    REQUIRE(copied.error().message() ==
            std::format("writing {}", (dst / "source.h.tmpl").string()));
}
